[Dark Mode] Update: Content Display Logic

yo1995 opened the pull request:
## Description

This PR adapts or updates every color in Content Display Logic folder to Dark Mode. It includes

- Change colors in `Main.storyboard` for each custom color in view controllers
- Add a color to `AppDelegate.swift`
- Change color for color picker view controller
- Change colors for the search bar, which is used both in content collection view controller (the main screen card view) and in `SearchForWebmapByKeywordViewController.swift`
- Change CSS styles for README and source code panes, so that they are responsive to Dark Mode
- Update colors in download progress view - for on-demand resource dependent samples

You can use <kbd>command</kbd> + <kbd>shift</kbd> + <kbd>A</kbd> to toggle Dark Mode in Simulator.

Also, I removed the force Light Mode flag, so there is no need to edit anything when testing! Reasons [below](https://github.com/Esri/arcgis-runtime-samples-ios/pull/899#discussion_r503593806).

## Does not change

- DemoTouchManager to show touches, no color is changed in this class.
- SVProgressHUD, it is a 3rd party dependency, and the author hasn't adapted it to iOS 13 yet. We consider replacing it with our own Swift implementation in total. It is used in many samples, including Trace utility network and Perform valve isolation trace.

## References

/common-samples/issues/1507
/common-samples/issues/2037

## Demo

![201012-demo](https://user-images.githubusercontent.com/9660181/95801035-e5403d00-0cad-11eb-9e05-0d25a23085b4.gif)

yo1995 reviewed (commented):
~~The first round of changes I made to the main display logic. After the meeting, I feel it would more approachable with smaller PRs that involves less changed files, so I'll discuss with Sarat and try to figure out the best approach.~~

outdated

yo1995 commented on arcgis-ios-sdk-samples/Content Display Logic/Controllers/SourceCodeViewController.swift:
```diff
@@ -38,16 +39,25 @@ class SourceCodeViewController: UIViewController, UIAdaptivePresentationControll
         self.webView = webView
         
-        if let filename = filenames.first {
+        if let filename = currentFilename {
+            loadHTMLPage(filename: filename)
+        } else if let filename = filenames.first {
+            currentFilename = filename
             loadHTMLPage(filename: filename)
         }
     }
     
+    /// Change the highlight.js rendered HTML when switch  between modes.
+    override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
+        super.traitCollectionDidChange(previousTraitCollection)
+        loadHTMLPage(filename: currentFilename)
+    }
+    
```
This ensures the css get loaded every time we swap the source files, so that the source code pane always be in correct color mode.

yo1995 commented on arcgis-ios-sdk-samples/Content Display Logic/Controllers/SampleInfoViewController.swift:
```diff
@@ -54,13 +54,15 @@ class SampleInfoViewController: UIViewController {
                 <link rel="stylesheet" href="\(cssPath)">
                 <link rel="stylesheet" type="text/css" href="https://cdnjs.cloudflare.com/ajax/libs/foundation/5.5.2/css/foundation.min.css">
                 <link rel="stylesheet" type="text/css" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.4.0/css/font-awesome.min.css">
+                <link rel="stylesheet" href="\(darkmodePath)">
```
Add an additional css to only adapt a few colors to dark mode automatically, to avoid changing the original css style.

yo1995 commented on arcgis-ios-sdk-samples/Content Display Logic/Storyboard/Main.storyboard:
```diff
@@ -267,17 +260,17 @@
+                                <rect key="frame" x="79.666666666666686" y="368" width="217" height="21"/>
                                 <fontDescription key="fontDescription" type="system" pointSize="17"/>
-                                <color key="textColor" red="0.66666666666666663" green="0.66666666666666663" blue="0.66666666666666663" alpha="1" colorSpace="custom" customColorSpace="sRGB"/>
+                                <color key="textColor" systemColor="secondaryLabelColor" red="0.23529411759999999" green="0.23529411759999999" blue="0.26274509800000001" alpha="0.59999999999999998" colorSpace="custom" customColorSpace="sRGB"/>
```
This is the detail label under a certain sample's cell in the list.

![image](https://user-images.githubusercontent.com/9660181/86976795-1c8fd280-c130-11ea-8137-4bc83ee48c12.png)

I decide to change it to a slightly different gray color, as it better fits the definition of `secondaryLabelColor`.

yo1995 commented on arcgis-ios-sdk-samples/Content Display Logic/Controllers/ListViewController.swift:
```diff
@@ -39,7 +39,6 @@ class ListViewController: UITableViewController {
         let cell = tableView.dequeueReusableCell(withIdentifier: "ListCell", for: indexPath)
         
         cell.textLabel?.text = list[indexPath.row]
-        cell.backgroundColor = .clear
```
No need to explicitly set this background color - bgc for cell is default to clear color.

yo1995 commented on arcgis-ios-sdk-samples/Content Display Logic/Controllers/SampleInfoViewController.swift:
```diff
@@ -24,14 +24,8 @@ class SampleInfoViewController: UIViewController {
-        // Prior to iOS 11, there was a bug in WKWebView.init(coder:) that
-        // caused a crash.
-        let webView = WKWebView(frame: view.bounds)
         webView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
         webView.navigationDelegate = self
```
Since we already dropped support for below iOS 12, there's no need to do this. I've added a WKWebView to main.storyboard and connected with an IBOutlet.

yo1995 commented on arcgis-ios-sdk-samples/Content Display Logic/Storyboard/Main.storyboard:
```diff
@@ -357,13 +350,31 @@
                         <rect key="frame" x="0.0" y="0.0" width="375" height="724"/>
                         <autoresizingMask key="autoresizingMask" widthSizable="YES" heightSizable="YES"/>
-                        <color key="backgroundColor" red="1" green="1" blue="1" alpha="1" colorSpace="custom" customColorSpace="sRGB"/>
+                        <subviews>
```
The IBOutlet of WKWebView added to storyboard.

yo1995 reviewed (commented):
~~This will be quite an effort to review and verify. Please take your time to build, look around and scrutinize these changes. 🙏 Later will check with Sarat to ensure they do not break anything.~~

outdated

yo1995 commented on arcgis-ios-sdk-samples/AppDelegate.swift:
```diff
@@ -169,9 +169,38 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UISplitViewControllerDele
-    class var backgroundGray: UIColor { return #colorLiteral(red: 0.973, green: 0.973, blue: 0.973, alpha: 1) }
-    class var primaryTextColor: UIColor { return #colorLiteral(red: 0.196, green: 0.196, blue: 0.196, alpha: 1) }
-    class var secondaryTextColor: UIColor { return #colorLiteral(red: 0.349, green: 0.349, blue: 0.349, alpha: 1) }
+    static let primaryBlue: UIColor = { return #colorLiteral(red: 0, green: 0.475, blue: 0.757, alpha: 1) }()
+    static let secondaryBlue: UIColor = { return #colorLiteral(red: 0, green: 0.368, blue: 0.584, alpha: 1) }()
+    
+    static let systemBackgroundWithOpacity: UIColor = {
+        if #available(iOS 13, *) {
+            return UIColor.systemBackground.withAlphaComponent(0.6)
+        } else {
+            return #colorLiteral(red: 1, green: 1, blue: 1, alpha: 0.6)
+        }
+    }()
+    
```
I'm not pretty sure about the changes here. I would like to make them `IBInspectable`, but using Color assets makes sense as well: https://nshipster.com/dark-mode/#manage-colors-with-an-asset-catalog.

Should I instead create these colors as asset catalog to avoid future re-visits?
Using asset catalog would also be beneficial for changing theme colors.

yo1995 replied:
Update: this change will get reverted and created as a color asset catalog.

philium commented on arcgis-ios-sdk-samples/Content Display Logic/Controllers/SourceCodeViewController.swift:
```diff
@@ -63,7 +73,12 @@ class SourceCodeViewController: UIViewController, UIAdaptivePresentationControll
-        let cssPath = Bundle.main.path(forResource: "xcode", ofType: "css") ?? ""
+        let cssPath: String
+        if traitCollection.userInterfaceStyle == .dark {
+            cssPath = Bundle.main.path(forResource: "solarized-dark", ofType: "css") ?? ""
+        } else {
+            cssPath = Bundle.main.path(forResource: "xcode", ofType: "css") ?? ""
```
Instead of specifying a default value in both places, you could make `cssPath` optional and then specify a default value at the point of use.

philium commented on arcgis-ios-sdk-samples/Content Display Logic/Views/DownloadProgressView.swift:
```diff
@@ -102,7 +102,7 @@ class DownloadProgressView: UIView {
         self.transparentShapeLayer.path = UIBezierPath(arcCenter: CGPoint(x: self.radius, y: self.radius), radius: self.radius, startAngle: 0, endAngle: 2 * CGFloat.pi, clockwise: true).cgPath
         self.transparentShapeLayer.fillColor = UIColor.clear.cgColor
-        self.transparentShapeLayer.strokeColor = UIColor(red: 0, green: 0, blue: 0, alpha: 0.3).cgColor
+        self.transparentShapeLayer.strokeColor = UIColor.white.cgColor
```
Does this mean the layer is no longer transparent? If so, should it be renamed to reflect that?

yo1995 commented:
> It seems a little odd that the navigation bar (other than the search text field, which isn't always visible) doesn't change in dark mode. I haven't looked at it yet, but I'm guessing we'll have separate branding colors in light and dark mode? Either way, I was thinking about the title text color. White in dark mode looks odd to me. Right now at least, because of the bar color, it is basically the inverse of `titleTextColor`. We could actually make it the inverse in dark mode as well by using a dynamic color:

My thoughts keep growing after discussions about UI color choices. 🤔 💭 

Esri does have a universal palette for different colors, but not for Dark Mode. I took an approach to pair up Dark Mode colors described here: https://github.com/Esri/arcgis-runtime-samples-ios/pull/915#discussion_r458290518

For the nav bar
- its background color should be different in light and dark. I've come up with a `primaryThemeColor` same as the logo color in #915 that both adapt light and dark mode
- title text color should always be white/light gray, given that the background color will be purple. I tried with black text in dark mode and it doesn't look good. I also tried other apps with theme color on my phone, such as Slack that uses a violet red theme - it keeps the title as light color both in light and dark mode.

Since Dark Mode is not as prioritized as App Icon and theme color update, we might consider push back this update and first go with #877 and #915. After setting up a universal color assets catalog, we can directly use named colors for convenience.

I'll convert this PR back to draft and revisit afterwards.

yo1995 commented on arcgis-ios-sdk-samples/Content Display Logic/Views/DownloadProgressView.xib:
```diff
@@ -33,21 +32,21 @@
                         </button>
                     </subviews>
-                    <color key="backgroundColor" white="1" alpha="1" colorSpace="calibratedWhite"/>
+                    <color key="backgroundColor" systemColor="secondarySystemBackgroundColor" red="0.94901960780000005" green="0.94901960780000005" blue="0.96862745100000003" alpha="1" colorSpace="custom" customColorSpace="sRGB"/>
```
Ahh it gonna be hard to resolve these merge conflicts! 😩

yo1995 reviewed (commented):
@vquach2404 @philium I've revamped this PR and make the content display logic Dark Mode compatible. Please give it a try! 🦉 

Thanks.

yo1995 commented on arcgis-ios-sdk-samples/Content Display Logic/Storyboard/Main.storyboard:
```diff
@@ -348,41 +344,69 @@
+                                    <constraint firstItem="KE4-Ic-Gt2" firstAttribute="top" secondItem="rA3-JQ-Mdx" secondAttribute="top" id="sqb-Ou-quV"/>
                                 </constraints>
+                            </view>
+                            <toolbar opaque="NO" clipsSubviews="YES" clearsContextBeforeDrawing="NO" contentMode="scaleToFill" translatesAutoresizingMaskIntoConstraints="NO" id="cZS-cc-Di0">
+                                <rect key="frame" x="0.0" y="764" width="414" height="44"/>
```
There is an issue with the toolbar bottom constraints right now. Since the Source Code View Controller is embedded in a container view, storyboard fails to snap the safe area with its superview.

![image](https://user-images.githubusercontent.com/9660181/95800689-da38dd00-0cac-11eb-9f02-3d6c7870b45e.png)

I've asked for help from our instructor in last training, but feel free to give me some help/thoughts!

philium replied:
I'm not seeing that with an iOS 14 simulator. Which simulator are you using?

![image](https://user-images.githubusercontent.com/2257493/96197231-47937a80-0f06-11eb-9aac-6eb3be96a802.png)

philium replied:
Never mind. I didn't see that some changes had been force pushed. After resetting my local branch, I now see the issue. I'll take a look.

yo1995 replied:
Sorry for the force push. There were quite a few conflicts and I decided to ditch the outdated changes and redo from scratch. 

The previous version set the container view's background to the same as the toolbar's bar tint to hide this problem, but that won't work under Dark Mode.

The workaround that I came up with is to 

1. Uncheck the translucent checkbox of toolbar, so that it doesn't have the UIVisualEffect view for translucency
2. Set its bar tint to the same as the base/root view's background color, such as `systemBackgroundColor`. 

However, toolbars in samples or in source code pane shouldn't be different in colors, so I didn't make this change.

yo1995 replied:
Solved by making a small tweak in c27369e 😏

yo1995 replied:
|Translucent|Opaque|
|-|-|
|![image](https://user-images.githubusercontent.com/9660181/96792919-363be980-13b0-11eb-80fc-ea0593f737c8.png)|![image](https://user-images.githubusercontent.com/9660181/96793046-70a58680-13b0-11eb-8d02-7a7a32c6c343.png)|

One more thing to mention: while the purple tint looks beautiful on an opaque bar, it might be a bit bright on a translucent bar. Should I tweak the accentColor for Dark Appearance?

philium replied:
Yeah, looks like that might be a good idea.

yo1995 replied:
Slightly adjusted the purple color in Dark Mode
|Before|After|
|-|-|
|![](https://user-images.githubusercontent.com/9660181/96792919-363be980-13b0-11eb-80fc-ea0593f737c8.png)|![image](https://user-images.githubusercontent.com/9660181/96924688-dd2f8c80-1467-11eb-864b-8767892e2525.png)|

vquach2404 reviewed (commented):
I tested it and it everything looks good besides the toolbar issue you mentioned. You may have more luck getting a response from our course leader if you direct message him or mention him in your message.

yo1995 commented:
@philium I tweaked the toolbar for source code pane and now it looks correct. 

In terms of changing the info button segue into present modally, I incline to stick to the current practice, i.e. tapping the info button push onto nav stack. 
<details>

There might be some samples that depends on the `viewDidAppear(_:)` method, and [iOS 13's interactive dismissal](https://developer.apple.com/documentation/uikit/uiviewcontroller/3229894-modalinpresentation) does not call it. Unless we make it a fullscreen modal presentation - but that doesn't make a huge difference to our current "show/push" approach.
<details>

philium commented on arcgis-ios-sdk-samples/AppDelegate.swift:
```diff
@@ -177,4 +177,6 @@ class AppDelegate: UIResponder, UIApplicationDelegate, UISplitViewControllerDele
 extension UIColor {
     // Also used as global tint/accent color.
     class var accentColor: UIColor { return UIColor(named: "AccentColor")! }
+    // The translucent background color for status labels.
+    class var statusLabelBackgroundColor: UIColor { return UIColor(named: "statusLabelBackgroundColor")! }
```
It looks like this property isn't being used. Am I missing something?

yo1995 replied:
I will need this color to replace (unify) many status label's background color in each category. I add it now so that we can branch from `v.next` to update Dark Mode for multiple categories in parallel, after this PR is merged.

If I add this color in a category's PR where it is first needed, it means we have to wait until that category's PR is merged before we can use this color in other categories.

What do you think?

philium replied:
I'm a big fan of not introducing code that isn't actually used yet. Plans change all the time and if the code ends up not getting used, it becomes a liability rather than an asset. However, I understand why you added this property now and I think that makes sense.

yo1995 replied:
Thanks!

yo1995 commented:
@saratchandrakarumuri We will continuously make progress to adapt to Dark Mode in the following couple of sprints! 

Please take a look at this PR to ensure the main screens (except sample-specific screens) are adapted correctly. Thanks! 🦉

saratchandrakarumuri commented:
* Download progress bar is not clearly visible in dark mode
![IMG_0C054881D576-1](https://user-images.githubusercontent.com/8692125/96949205-31546400-149c-11eb-98f4-dd22f926f26c.jpeg)

* In full screen mode, current dark mode looks good, but on an iPad in split mode, colors look little clumsy with multiple shades of dark

![IMG_0119](https://user-images.githubusercontent.com/8692125/96949397-a58f0780-149c-11eb-9e4e-a419967e3ddb.PNG)
![IMG_0120](https://user-images.githubusercontent.com/8692125/96949406-a7f16180-149c-11eb-8755-92bf6111c82f.PNG)

yo1995 commented:
@saratchandrakarumuri 

* Download progress bar is not clearly visible in dark mode

![progressView](https://user-images.githubusercontent.com/9660181/97060768-489d5b00-1549-11eb-9b2b-db93974d7114.gif)

I tried multiple pairs of grayscale colors with no luck. 🤕 None of them look perfect in both Light and Dark Mode. 
So this time I replaced the progressing stroke with green. Does using green to represent progress make sense? @philium I thought to use the purple accent color for progress, but I remember that color should only used by interact-able elements on the screen.

* In full screen mode, current dark mode looks good, but on an iPad in split mode, colors look little clumsy with multiple shades of dark

Fixed.

philium commented:
> Does using green to represent progress make sense?

I think the green looks fine.

yo1995 commented:
@saratchandrakarumuri This is ready for another look. 👀

saratchandrakarumuri commented:
@yo1995 Sorry forgot to submit the comment 🙈 ...
* Download progress bar looks good
* I don't know the design decisions and effort behind it, but even though now the while viewer is same color I feel grey color is easier on eyes than solid black... have you tried using grey everywhere?

yo1995 commented:
> I don't know the design decisions and effort behind it, but even though now the while viewer is same color I feel grey color is easier on eyes than solid black... have you tried using grey everywhere?

This makes sense. In fact some apps does use [`secondarySystemBackground`](https://developer.apple.com/documentation/uikit/uicolor/3173137-secondarysystembackground) (dark gray) as the major dark bgc. Currently for most places I just used the default color recommended by Apple. @philium What's your suggestion?
I'll make another commit to switch gray and give it a try, and we can go back if it doesn't look well.

yo1995 commented:
@saratchandrakarumuri I've changed some of the solid black colors into dark gray, and it does look softer. 

@vquach2404 @philium How do you like it comparing to the previous full black `#000000`? 
I used the `tertiarySystemBackgroundColor` (`#FFFFFF`,
`#2C2C2E`) for the main collection view, version info VC as well as source code and README background color.

![Simulator Screen Shot - iPhone 11 - 2020-10-28 at 14 55 13](https://user-images.githubusercontent.com/9660181/97501182-1e271580-192e-11eb-96d9-8c46c5c9a6b6.png)

If we don't want this change, it is easy to just revert 4ad2c34 to go back.

vquach2404 commented:
I think the new gray color looks great! I noticed that the list of samples is still black. Will that be changing to gray as well?

yo1995 commented:
> I think the new gray color looks great! I noticed that the list of samples is still black. Will that be changing to gray as well?

I left that as solid black `#000000` on purpose. 
We uses lots of tableviews in our app, and I tend to set all of them with a single color/style, so I didn't change its default settings (which is black by default and dark gray when row selected).

But it is very easy to set it to gray. Wait for more input from Phil and Sarat.

saratchandrakarumuri commented:
@yo1995 If we change that one to grey like the rest of the viewer, what would be the selection color?

vquach2404 reviewed (commented):
![Simulator Screen Shot - iPhone 12 Pro Max - 2020-10-29 at 16 53 59](https://user-images.githubusercontent.com/52215269/97645387-18066700-1a0a-11eb-9037-87e85071d3fa.png)
This looks better! My concern is about the color of buttons. Although the button should be a different color than the background, the black against the gray seems slightly out of place. Could we make buttons just a darker gray?

yo1995 commented:
> This looks better! My concern is about the color of buttons. Although the button should be a different color than the background, the black against the gray seems slightly out of place. Could we make buttons just a darker gray?

@vquach2404  Do you mean making the toolbar's background a gray color, instead of solid black? That is quite hard to achieve. 🤔 

![view hierarchy](https://user-images.githubusercontent.com/9660181/97746662-f956ad80-1aa7-11eb-98bf-879e8366ed1c.png)

The toolbar by default does not have a background color; instead, it is "translucent" like a frosted glass. So the color is actually a Gaussian blur of the layers beneath. See the view hierarchy above.

The reason why it is solid black is that currently the Apple default background color for the root views for each sample (as well as the README and source code panes) is `systemBackground` - which is solid black and solid white.

![gray-toolbar](https://user-images.githubusercontent.com/9660181/97747281-f27c6a80-1aa8-11eb-99ac-0187d0a199a6.png)

After changing the root view's bgc to gray it looks like above. It is undesirable to have toolbars with inconsistent colors. We can go with 1 of the 3 options below.

- Keep it as is. 
- Override the toolbar background color in AppDelegate. 
- Changing each sample's root view's bgc to `tertiarySystemBackground` so that toolbar blurs to dark gray.

I tend to keep it as is, so that we can always use the default color whenever we add new samples or edit existing ones.

vquach2404 commented:
> * Override the toolbar background color in AppDelegate.
I personally find the drastic contrast of the toolbar against the background odd so this option seems best to me. Does anyone else feel the same @saratchandrakarumuri  @philium ? If not, we'll keep the toolbars the way it is.

philium commented:
I too think the toolbar looks odd showing up as black. If we could make it gray somehow, that would probably be better.

yo1995 commented:
Ready for another review. If we are happy with the new colors, we'll proceed to update all categories in the following days before release.

- Fix the colors in samples for Dark Mode
- Update the screenshots to reflect new colors
- Fix any wrong constraints in the storyboard that affects appearance

Details at: https://github.com/Esri/arcgis-runtime-samples-ios/pull/915#issuecomment-692327353

saratchandrakarumuri reviewed (approved):
New look is good and consistent

vquach2404 reviewed (approved):
Thanks for making that change, it looks good!

yo1995 commented:
I'm very sorry to find that I got the background color name wrong in a3bf026b 🤦‍♂️🤦‍♂️

```diff
- statusLabelBackgroudColor
+ statusLabelBackgroundColor
```

Will log an issue for it...
